benchmarks/advection_reaction_3D/raja/advection_reaction_3D.cpp

```diff
@@ -159,182 +154,121 @@ UserData::~UserData()
-      return -1;
-    }
+  if (wmask != nullptr) {
+    N_VDestroy(wmask);
+    wmask = nullptr;
   }
 
-  return(0);
+  /* free the parallel grid */
+  delete grid;
 }
 
 
-/* Starts the exchange of the neighbor information */
-int ExchangeAllStart(N_Vector y, UserData* udata)
+/* --------------------------------------------------------------
+ * Communication functions
+ * --------------------------------------------------------------*/
+
+/* Fills send buffers before exchanging neighbor information */
+int FillSendBuffers(N_Vector y, UserData* udata)
 {
-  SUNDIALS_MARK_BEGIN(udata->prof, "Neighbor Exchange");
 
   /* shortcuts */
-  realtype c = udata->c;
+  const realtype c = udata->c;
+  const int nxl = udata->grid->nxl;
+  const int nyl = udata->grid->nyl;
+  const int nzl = udata->grid->nzl;
+  const int dof = udata->grid->dof;
 
-  /* extract the data */
+  /* Create a 4D view of the vector */
   RAJA::View<realtype, RAJA::Layout<NDIMS+1> > Yview(GetVecData(y),
-                                                     udata->grid->nxl,
-                                                     udata->grid->nyl,
-                                                     udata->grid->nzl,
-                                                     udata->grid->dof);
+                                                     nxl, nyl, nzl, dof);
 
   if (c > 0.0)
   {
+
     /* Flow moving in the positive directions uses backward difference. */
-    udata->grid->ExchangeStart(
-      [=] (realtype*, realtype* Esend, realtype*, realtype* Nsend, realtype* Bsend, realtype*) {
-        int nxl = udata->grid->nxl;
-        int nyl = udata->grid->nyl;
-        int nzl = udata->grid->nzl;
-        int dof = udata->grid->dof;
-
-        auto range = RAJA::make_tuple(RAJA::RangeSegment(0, std::max(1,nxl-1)),
-                                      RAJA::RangeSegment(0, std::max(1,nyl-1)),
-                                      RAJA::RangeSegment(0, std::max(1,nzl-1)));
-
-        RAJA::View<realtype, RAJA::Layout<3> >
-          Eview(Esend, nyl, nzl, dof);
-        RAJA::View<realtype, RAJA::Layout<3> >
-          Nview(Nsend, nxl, nzl, dof);
-        RAJA::View<realtype, RAJA::Layout<3> >
-          Bview(Bsend, nxl, nyl, dof);
-
-        RAJA::kernel<XYZ_KERNEL_POL>(range,
-          [=] DEVICE_FUNC (int i, int j, int k) {
-
-          if (nxl > 1)
-          {
-            Eview(j,k,0) = Yview(nxl-1,j,k,0);
-            Eview(j,k,1) = Yview(nxl-1,j,k,1);
-            Eview(j,k,2) = Yview(nxl-1,j,k,2);
-          }
-
-          if (nyl > 1)
-          {
-            Nview(i,k,0) = Yview(i,nyl-1,k,0);
-            Nview(i,k,1) = Yview(i,nyl-1,k,1);
-            Nview(i,k,2) = Yview(i,nyl-1,k,2);
-          }
-
-          if (nzl > 1)
-          {
-            Bview(i,j,0) = Yview(i,j,nzl-1,0);
-            Bview(i,j,1) = Yview(i,j,nzl-1,1);
-            Bview(i,j,2) = Yview(i,j,nzl-1,2);
-          }
-
-        });
-      });
+
+    /* Create 3D views of send buffers */
```

Comment:
Should this check the values of `NDIMS` for which buffers need to be packed?

Reply:
Good catch.  This one is an easy fix.